api/anchor.capnp

```diff
@@ -24,8 +24,86 @@ interface Anchor {
     # anchors in the path that do not currently exist are created along
     # the way.
 
+    cell @2 () -> (loader :Loader, storer :Storer, swapper :Swapper);
+    # Cell returns a set of capabilities that provide access to a boxed
+    # value, assigned to the Anchor. Zero or more of these capabilities
+    # MAY be witheld, in which case the corresponding return value will
+    # be null. The loader and storer capabilities respectively map onto
+    # read and write permissions. The swapper is an extended read/write
+    # capability, which SHALL be witheld if either loader or storer are
+    # witheld.
+
     struct Child {
         anchor @0 :Anchor;
         name   @1 :Text;
     }
+
+    struct Value {
```

Comment:
Yeah, I'm wondering the same thing.

My gut tells me that we want to discourage people from overusing anchors, since shared memory is tricky to get right in a concurrent/distributed context, but obviously we don't want to be overly restrictive either.  It's not immediately obvious to me what the optimum is, here.  As a result, I'd like to punt on that decision until we have some more clarity and experience using anchors in real applications.  Fortunately, capnp lets us add fields to union types while preserving backwrards-compatibily.

This having been said, I'm very interested in your thoughts on this, as well as your feedback when you start to implement the demo application.  I see roughly three approaches we could take:

1. "CSP Only".  The only things you can load/store/swap from an anchor are `Proc`s and `Chan`s.  This maximizes safety for the user by effectively forcing them into the CSP paradigm, but also makes it awkward to do simple things like load a piece of config (which would require a dedicated process+channel pair).
2. "Capabilities Only".  Any capability can be loaded/stored/swapped.  This would involve extending the `union` with a `Capability` type.  Capnp provides some ordering guarantees on method calls, which might help users design good APIs for dealing with stateful data.  This is a bit more of a "safeties off" design.
3.  "Anything goes".  We add an `AnyPointer` field to the `union` and call it a day.  Alternatively, we could even forego the union entirely in favor of `AnyPointer` everywhere (which is roughly analogous to `interface{}` in Go).  It's not clear to me how much riskier this is relative to a capabilities-only design.

A related, though orthogonal, concern is that we lose typing information in options 2 & 3.  One thing we could do is to allow users to define a schema on an anchor, so that others know what kind of data they are dealing with.  This could be done quite straightforwardly using Capnp's schema definition (which, conveniently, is itself a capnp schema!), or we could define our own type system in capnp.  Both have advantages.

Again, I very enthusiastically welcome any discussion or brainstorming around this issue.  I'm always happy to hop on a call and just talk through various ideas, because there's a lot to unpack here.  In the meantime, I think we should resist the urge to solve everything at once, moving slowly and deliberately instead.